### PR TITLE
Add OpenTracing support for Go streaming RPCs.

### PR DESCRIPTION
**Pull request by rnburn**
This PR adds additional interceptors to `otgrpc` so that streaming RPCs can be traced. 

One of the more challenging problems with streaming RPCs is figuring out when to finish the span on the client-side. Unlike other interceptors, gRPC's client-side stream interceptor functions don't enclose the entire RPC, but rather return a [ClientStream](https://godoc.org/google.golang.org/grpc#ClientStream) that the invoker then interacts with. Based on the error codes returned by ClientStream's methods you can figure out in most use-cases when the span finishes, though there's special cases that are problematic ([example](http://stackoverflow.com/q/42915337/4447365)).

I added code for the problematic cases that at least eventually finish the spans and clean up resources. And I reached out on [StackOverflow](http://stackoverflow.com/q/42988396/4447365) to see if to see if there's any other solution, but I'm not sure it's possible with the current API to do any better.

I also added testing coverage that follows the structure of this [test](https://github.com/grpc/grpc-go/blob/master/test/end2end_test.go) in gRPC Go.

### CODE REVIEW

**Comment by tedsuo** [comment]
LGTM 👍

**Comment by danielamiao** [comment]
From a coding/implementation perspective, this all looks good to me, but I’m concerned about the general idea/design of a single `span` around a streaming client or server, which can last up to hours. This definitely breaks the definition of a `span`, so I think it's worthwhile to step back and discuss the general paradigm of instrumenting a streaming RPC.

Some quick thoughts (all IMHO):
1. The top `span` of a streaming RPC should be finished when the stream is successfully established, this applies to both client and server streams.
2. Each subsequent `message` sent on the above stream should have its own `span`, finishing when the `SendMsg` call finishes (could be an error).
3. Similarly, each `RecvMsg` call should have its own span.
4. For parent-child relationships, one way is to make all spans of messages on the stream children of the parent top `span` in 1), but that'll likely create huge long traces, so I don't actually think we should do it.

**Comment by rnburn** [comment]
I agree that tracing individual requests in the stream would be useful, but gRPC only allows you to send metadata when you initiate the stream, so I don’t see how you could propagate context if you created a span per-request. 

I can see creating a span for the entire lifetime of the stream as being useful for shorter lived streams. I’m not sure what you could do for long-lived streams without being able to send context for individual requests.

File: go/otgrpc/client.go [hunk 1]
```diff
@@ -81,3 +72,167 @@ func OpenTracingClientInterceptor(tracer opentracing.Tracer, optFuncs ...Option)
 		return err
 	}
 }
+
+// OpenTracingStreamClientInterceptor returns a grpc.StreamClientInterceptor suitable
```
Comment (bhs) [inline go/otgrpc/client.go:76]: Please clarify that the instrumentation creates a single Span for the entire streaming request (not one Span per streaming response).

File: go/otgrpc/server.go [hunk 2]
```diff
@@ -65,11 +61,80 @@ func OpenTracingServerInterceptor(tracer opentracing.Tracer, optFuncs ...Option)
...
+			serverSpan.LogFields(log.String("event", "error"), log.String("message", err.Error()))
 		}
 		if otgrpcOpts.decorator != nil {
 			otgrpcOpts.decorator(serverSpan, info.FullMethod, req, resp, err)
 		}
 		return resp, err
 	}
 }
+
+// OpenTracingStreamServerInterceptor returns a grpc.StreamServerInterceptor suitable
```
Comment (bhs) [inline go/otgrpc/server.go:73]: (same comment about clarifying how many Spans will be used to model the streaming RPC)

File: go/otgrpc/server.go [hunk 3]
```diff
@@ -65,11 +61,80 @@ func OpenTracingServerInterceptor(tracer opentracing.Tracer, optFuncs ...Option)
...
+			info.FullMethod,
+			ext.RPCServerOption(spanContext),
+			gRPCComponentTag,
+		)
+		defer serverSpan.Finish()
+		ss = &openTracingServerStream{
+			ServerStream: ss,
+			ctx:          opentracing.ContextWithSpan(ss.Context(), serverSpan),
+		}
+		err = handler(srv, ss)
```
Comment (bhs) [inline go/otgrpc/server.go:114]: so there's no way to get callbacks about intermediate responses, correct? sigh...

**Review by bhs — COMMENTED** [review]
(sorry for the insane delay)

File: go/otgrpc/client.go (hunk 1 above)
Comment (rnburn) [reply go/otgrpc/client.go:76]: Added clarifying comments.

File: go/otgrpc/server.go (hunk 3 above)
Comment (rnburn) [reply go/otgrpc/server.go:114]: The stream interceptor doesn't provide a callback interface, but you can still intercept them by customizing the ServerStream's SendMsg and RecvMsg functions similar to what's done with the [ClientStream](https://github.com/rnburn/grpc-opentracing-1/blob/go-stream-rpc/go/otgrpc/client.go#L202).

It is possible to get callback-style interception via the `stats.Handler` interface, and @charithe has written an OpenTracing [implementation](https://github.com/charithe/otgrpc) using that approach; though that would be less consistent with how OpenTracing is instrumented elsewhere in gRPC.

File: go/otgrpc/client.go [hunk 4]
```diff
@@ -81,3 +72,168 @@ func OpenTracingClientInterceptor(tracer opentracing.Tracer, optFuncs ...Option)
...
+	finishChan := make(chan struct{})
+
+	isFinished := new(int32)
+	*isFinished = 0
+	finishFunc := func(err error) {
+		// The current OpenTracing specification forbids finishing a span more than
+		// once. Since we have multiple code paths that could concurrently call
+		// `finishFunc`, we need to add some sort of synchronization to guard against
+		// multiple finishing.
+		if !atomic.CompareAndSwapInt32(isFinished, 0, 1) {
```
Comment (tamird) [inline go/otgrpc/client.go:140]: why can't this all use a `sync.Once`? The case of `finishedChan` being closed without `<-cs.Context().Done()` being closed is exactly the case where the finalizer has run, so there's no need for this closure to know about `finishedChan` at all, and there's no need for this atomic (use sync.Once instead). What do you think?

Comment (rnburn) [reply go/otgrpc/client.go:140]: Not sure I follow: The finalizer calls this same closure. If the closure doesn't know about `finishedChan`, how does the channel get closed and the Go-routine not leaked?

Comment (tamird) [reply go/otgrpc/client.go:140]: Just close it right in the finalizer, it's the only place that needs to bother.

Comment (rnburn) [reply go/otgrpc/client.go:140]: I'm open to doing that. But I'm not sure that makes the code any simpler: it's just moving the closing from one closure to another, no?

And just to be clear on the finalizer, it's only meant to handle the special case when the stream isn't read until [`io.EOF`](https://stackoverflow.com/questions/42915337/are-you-required-to-call-recv-until-you-get-io-eof-when-interacting-with-grpc-cl/42939914#42939914). I would expect the most common usage to be that `<- cs.Context().Done()` isn't closed and the finished is done by intercepting an [`io.EOF`](https://github.com/rnburn/grpc-opentracing-go/blob/master/go/otgrpc/client.go#L204).

Comment (tamird) [reply go/otgrpc/client.go:140]: Ah, I see! I must have misread the first time through; apologies. I still think you can substitute the atomic for a `sync.Once` to get slightly more expected (to me) behaviour where no calls to `finishFunc` will return until the finishing is actually complete.

Comment (rnburn) [reply go/otgrpc/client.go:140]: I could see how that behavior could be preferred in most situations, but is it really the best fit for tracing?

Given that the tracing code should be orthogonal to the rest of the functioning of an app, I would think you'd want it to get out of the way as much as possible. For example, if one thread cancels the RPC call via the context while another thread reads an `io.EOF` from `RecvMsg`, do you really want one of the code paths to take longer just to wait on the span's finishing code to complete?

Comment (tamird) [reply go/otgrpc/client.go:140]: I could argue either way, so I won't.

**Comment by bcotton** [comment]
Any motion on this PR? It would be nice to have in the official release.

Thanks

**Comment by trusch** [comment]
Any motion on this would indeed be nice. From my point of view tracing complete streams instead of every message is better than not having tracing support for streaming rpc calls at all.

My personal problem is that the master of moby/buildkit depends on that PR, and that breaks stuff on my side.

Thanks in advance!

**Comment by tedsuo** [comment]
Merging this in. Sorry for the wait!

FYI: this repo is about to be moved to https://github.com/opentracing-contrib/go-grpc, and future development will happen there, so that we can monitor PR activity better and with more accountability.

**Pull request merged**